cdmInstance may not be available when webKitMediaOpenCDMDecryptorRese…

goruklu opened the pull request:
…tSessionFromKeyIdIfNeeded() is called

This prevents a crash seen when playing content in http://www.gurdal-lab.net/dash.js/samples/playreadyMS.html

Signed-off-by: Gurdal Oruklu <gurdal_oruklu@comcast.com>

goruklu committed:
cdmInstance may not be available when webKitMediaOpenCDMDecryptorResetSessionFromKeyIdIfNeeded() is called

This prevents a crash seen when playing content in http://www.gurdal-lab.net/dash.js/samples/playreadyMS.html

Signed-off-by: Gurdal Oruklu <gurdal_oruklu@comcast.com>

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -153,6 +153,12 @@ static SessionResult webKitMediaOpenCDMDecryptorResetSessionFromKeyIdIfNeeded(We
     RefPtr<WebCore::CDMInstance> cdmInstance = webKitMediaCommonEncryptionDecryptCDMInstance(self);
     ASSERT(cdmInstance && is<WebCore::CDMInstanceOpenCDM>(*cdmInstance));
+
+    if(!cdmInstance){
```
There is an assert before this, so please, let's adapt this into `if (!cdmInstance || !is<WebCore::CDMInstanceOpenCDM>(*cdmInstance))`. Then remove the assert before and move it to inside the if in form of `ASSERT_NOT_REACHED()`.

eocanha replied:
Before accepting this patch or its corrected version, I'm trying to understand why the CDMInstance isn't available on time.

calvaris committed:
[GStreamer][EME] Lock while CDM updates keys

This way we can prevent part of the race conditions happening while an
updateLicense was issued but not completed and the decryptors asking
for key ids.

goruklu committed:
moved the assert for cdmInstance check in webKitMediaOpenCDMDecryptorResetSessionFromKeyIdIfNeeded()

Signed-off-by: Gurdal Oruklu <gurdal_oruklu@comcast.com>

goruklu commented:
@calvaris I've updated the ASSERT() as you requested. Let me know if this looks OK to you. I'm still seeing this crash even after the change which locks CDM key updates.

calvaris commented on Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp:
```diff
@@ -152,9 +152,9 @@ static SessionResult webKitMediaOpenCDMDecryptorResetSessionFromKeyIdIfNeeded(We
 
-    if(!cdmInstance){
+    if( !cdmInstance || !is<WebCore::CDMInstanceOpenCDM>(*cdmInstance) )
+        ASSERT_NOT_REACHED();
```
We should ASSERT always after gstreamer logging, that way the log would be the last thing we see before the crash in debug mode.

calvaris reviewed (changes requested):
Besides this you need to remove my commit from this PR. My commit has still to be investigated and it corresponds to fixing or mitigating another problem, not this one.

This one is weird. It looks like the app is not attanching the MediaKeys in time and we need to know why.

goruklu commented:
replaced with #668
